Address issue 611: acknowledge dropped puts due to bucket type mismatches

bowrocker opened the pull request:
The repl bucket types implementation is failing to acknowledge puts that are dropped on the sink, which leaves the dropped item in the RTQ, which will eventually fill. 
- Make sure to call DoneFun() in the mismatch case, so the drop will be acknowledged and cleared.
- Make logging of the mismatch a debug.
- Make the logging report the offending bucket type.
- Add a counter of mismatch events and fire a timer to report it. 
- Add a configuration parameter to control the interval of reporting: bucket_type_drop_report_interval, default 10s.

Issue: https://github.com/basho/riak_repl/issues/611
Riak test: https://github.com/basho/riak_test/pull/676
### Review checklist
- [ ] manual verification of code
- [ ] eunit (w/ gist of output)
- [ ] EQC (w/ gist of output)
- [ ] riak_test (w/ gist of output)
- [ ] Dialyzer
- [ ] XRef
- [ ] Coverage reports
- [ ] internal docs (design docs)
- [ ] man pages

bowrocker committed:
Fixes for issue 611.

- Call DoneFun() in the error case during bucket type matching.
- Add a per bucket-type counter of drops due to type mis-match.
- Add a counter of drops and report them after a timer and then clear.

bowrocker committed:
Change bucket type drop reporting interval to 10s.

bowrocker committed:
Fix dialyzer failure in report_bt_drops.

jonmeredith commented on src/riak_repl2_rtsink_conn.erl:
```diff
@@ -217,7 +229,12 @@ handle_info(reactivate_socket, State = #state{remote = Remote, transport = T, so
-    end.
+    end;
+handle_info(report_bt_drops, State=#state{bt_drops = DropDict}) ->
+    _ = [ lager:debug("drops due to missing or mismatched type ~p:~p", 
```
Now that we're rate limited, this can be an error message. Would be good to give a count so operators can know how many updates were lost to help plan any mitigations after discovered.  `dict:fold` would give you both easily.

jonmeredith commented on src/riak_repl2_rtsink_conn.erl:
```diff
@@ -217,7 +229,12 @@ handle_info(reactivate_socket, State = #state{remote = Remote, transport = T, so
+    _ = [ lager:debug("drops due to missing or mismatched type ~p:~p", 
+                  [BucketType, dict:fetch(BucketType, DropDict)]) ||
+        BucketType <- dict:fetch_keys(DropDict) ],
+    {noreply, State#state{bt_drops = dict:new(), bt_timer = undefined}}.
```
May be nice to recheck interval from the appenv here so you can adjust without having to restart the connection.

jonmeredith commented:
Very quick scan of the code looks good.  Looks like calling the DoneFun on drops is valid, and that cascaded realtime would still work if you had A <-> B <-> C configured with A and C having matching types and B misconfigured, data would still arrive at B (although still would not fullsync).

Over time we may need to support other reasons for drops than bucket type misconfigurations (perhaps security restrictions), but we can refactor the reporting/state names then as it is all process-local.

I won't have time to complete review/running tests, so please could somebody else take the official review.

bowrocker committed:
Address review feedback for issue 611.

- Report total count of drops.
- Check app-config value of bt drop interval during reporting for
  updating.

bowrocker committed:
Clean up for issue 661.

bowrocker committed:
Change default bucket type drops reporting interval back to 10s.

kellymclaughlin commented:
Shouldn't this be opened against the `2.0` branch?

bowrocker commented:
Closing in favor of https://github.com/basho/riak_repl/pull/613.
